oauth2_provider/views/mixins.py

```diff
@@ -245,3 +254,29 @@ def get_scopes(self, *args, **kwargs):
 
         # this returns a copy so that self.required_scopes is not modified
         return scopes + [self.read_write_scope]
+
+class ClientProtectedResourceMixin(OAuthLibMixin):
+
+    """Mixin for protecting resources with client authentication as mentioned in rfc:`3.2.1`
+    This involves authenticating with any of: HTTP Basic Auth, Client Credentials and Access token in that order.
+    Breaks off after first validation. 
+    """
+
+    def dispatch(self, request, *args, **kwargs):
+        # let preflight OPTIONS requests pass
+        if request.method.upper() == "OPTIONS":
+            return super().dispatch(request, *args, **kwargs)
+        # Validate either with HTTP basic or client creds in request body.
+        # TODO: Restrict to POST.
+        valid = self.authenticate_client(request)
```

Comment:
I feel like this should also get `valid, r` where `r.user` is user associated with authenticated client. This would allow introspection to verify if given client is authorized to introspect given token. Can someone verify if this is the correct approach?